### DISCUSSION TITLE
Maintain focus on trigger when Menu is open

### DISCUSSION THREAD

**Discussion by ritz078** [adobe/react-spectrum#4458] [Q&A] [answered]
I am trying to implement a Toolbar with Menu using react-aria where you can have multiple menus and you can quickly navigate through them using right/left arrows. Something similar to https://www.radix-ui.com/docs/primitives/components/menubar where the focus stays on the trigger even after open the menu. The focus only shifts after we click on the up/down arrow keys.

Here's what I have so far:

https://user-images.githubusercontent.com/5389035/235486013-432ec141-a901-417a-bc17-34324b6be87a.mp4

As you can see, I am using [useMenuTriggerState](https://react-spectrum.adobe.com/react-stately/useMenuTriggerState.html)'s `state.open("first")` to open the menu when focus comes back to the menu. Even if I use `state.open()`, the focus shifts to the menu content and I am not able to bring it back to the trigger button without closing the menu content.

**Comment by snowystinger** [comment]
I don't think we've implemented this pattern yet, it's a menubar https://www.w3.org/WAI/ARIA/apg/patterns/menubar/examples/menubar-editor/

Menu bars still use menus inside them, so you could probably get most of it implemented using our hooks, but some things you'd need to handle yourself.

Could open a feature request for the menubar in React Aria and given that the behaviors and dom structure are well defined, you could contribute that if you wanted. Keep in mind that due to ongoing team priorities we won't have much time to actively support a contribution of this scale, but we'd appreciate any findings and progress you make here!

**Reply by ritz078** [reply]
Thank you. I have achieved most of the menubar pattern using react-aria (as you can see in the video). The only thing I am not able achieve is preventing autofocus of content when the menubar opens. 😞 That feature will enable the development of menubar.

Maybe a `preventContentAutoFocus` in `useMenu` will help here. The focus can shift when someone presses keyboard up/down arrows.

**Reply by snowystinger** [reply]
Yeah, that's the one place where I am not sure the best way to go about it, given that popovers generally need to contain focus. I'll bring it up with the core team at our next meeting as well as with our accessibility team. I don't think useMenu is doing anything with the focus though. That'd be useMenuTrigger most likely.

**Reply by ritz078** [reply, marked as answer]
Yes I think it's the popover that has focus. I realised that I can actually fake this behaviour by pretending that the menu trigger button is focused using CSS. So even though it's not focused I can style it to look like it does and add interactions using JS.


https://user-images.githubusercontent.com/5389035/235519721-9e1fdc81-beaf-4075-a8a6-1eb1487561b2.mp4

Thanks a lot for your help. I think I have achieved the most of what I wanted using existing APIs of react-aria.

**Reply by snowystinger** [reply]
That looks pretty good. I think that's an appropriate way to solve it until it can really be addressed in react-aria. Thanks for sticking with it!

**Reply by ritz078** [reply]
Thanks for all your help.

**Reply by BMCwebdev** [reply]
Hey everyone, jumping into this old thread with a quick shoutout.

I know there's already a marked answer here (and the CSS/JS workaround still holds up), but that was ~3 years ago and I'm curious where things stand today.

**Is React Aria considering first-class support for the [W3C Menubar pattern](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/)?** Or is it still the recommended approach to roll your own using the existing Menu / MenuTrigger / Popover hooks or components?

If anyone in the community has implemented this with what's currently available in React Aria, I'd love to hear how you approached it.

Thanks!
